src/System.IO.FileSystem/src/System/IO/DosMatcher.cs

```diff
@@ -0,0 +1,314 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Text;
+
+namespace System.IO
+{
+    internal static class DosMatcher
+    {
+        // [MS - FSA] 2.1.4.4 Algorithm for Determining if a FileName Is in an Expression
+        // https://msdn.microsoft.com/en-us/library/ff469270.aspx
+        private static readonly char[] s_wildcardChars =
+        {
+            '\"', '<', '>', '*', '?'
+        };
+
+        /// <summary>
+        /// Change '*' and '?' to '&lt;', '&gt;' and '"' to match Win32 behavior. For compatibility, Windows
+        /// changes some wildcards to provide a closer match to historical DOS 8.3 filename matching.
+        /// </summary>
+        public unsafe static string TranslateExpression(string expression)
+        {
+            if (string.IsNullOrEmpty(expression) || expression == "*.*")
+                return "*";
+
+            StringBuilder sb = StringBuilderCache.Acquire(expression.Length);
+            int length = expression.Length;
+            fixed (char* c = expression)
+            {
+                for (int i = 0; i < length; i++)
+                {
+                    switch (c[i])
```

Comment:
Why are we using unsafe code here?  I would expect the JIT would do a very good job with:
```C#
foreach (char c in expression)
```
or
```C#
for (int i = 0; i < expression.Length; i++)
{
    char c = expression[i];
    ...
}
```

Reply:
Fair enough- I'll trust the JIT :)